probes/hasDangerousWorkflowScriptInjection/patch/impl.go

```diff
@@ -0,0 +1,580 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package patch
+
+import (
+	"fmt"
+	"regexp"
+	"slices"
+	"strings"
+
+	"github.com/go-git/go-billy/v5/memfs"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/storage/memory"
+	"github.com/rhysd/actionlint"
+
+	"github.com/ossf/scorecard/v5/checker"
+	"github.com/ossf/scorecard/v5/checks/fileparser"
+	sce "github.com/ossf/scorecard/v5/errors"
+)
+
+type unsafePattern struct {
+	idRegex      *regexp.Regexp
+	replaceRegex *regexp.Regexp
+	envvarName   string
+	ghVarName    string
+}
+
+// Fixes the script injection identified by the finding and returns a unified diff users can apply (with `git apply` or
+// `patch`) to fix the workflow themselves. Should an error occur, an empty patch is returned.
+func GeneratePatch(
+	f checker.File,
+	content string,
+	workflow *actionlint.Workflow,
+	workflowErrs []*actionlint.Error,
+) (string, error) {
+	patchedWorkflow, err := patchWorkflow(f, content, workflow)
+	if err != nil {
+		return "", err
+	}
+	errs := validatePatchedWorkflow(patchedWorkflow, workflowErrs)
+	if len(errs) > 0 {
+		return "", fileparser.FormatActionlintError(errs)
+	}
+	return getDiff(f.Path, content, patchedWorkflow)
+}
+
+// Returns a patched version of the workflow without the script injection finding.
+func patchWorkflow(f checker.File, content string, workflow *actionlint.Workflow) (string, error) {
+	unsafeVar := strings.Trim(f.Snippet, " ")
+	runCmdIndex := f.Offset - 1
+
+	lines := strings.Split(content, "\n")
+
+	unsafePattern, err := getUnsafePattern(unsafeVar)
+	if err != nil {
+		return "", err
+	}
+
+	existingEnvvars := parseExistingEnvvars(workflow)
+	unsafePattern, err = useExistingEnvvars(unsafePattern, existingEnvvars, unsafeVar)
+	if err != nil {
+		return "", err
+	}
+
+	replaceUnsafeVarWithEnvvar(lines, unsafePattern, runCmdIndex)
+
+	lines, err = addEnvvarToGlobalEnv(lines, existingEnvvars, unsafePattern, unsafeVar)
+	if err != nil {
+		return "", sce.WithMessage(sce.ErrScorecardInternal,
+			fmt.Sprintf("Unknown dangerous variable: %s", unsafeVar))
+	}
+
+	return strings.Join(lines, "\n"), nil
+}
+
+func getUnsafePattern(unsafeVar string) (unsafePattern, error) {
+	unsafePatterns := []unsafePattern{
+		newUnsafePattern("AUTHOR_EMAIL", `github\.event\.commits.*?\.author\.email`),
+		newUnsafePattern("AUTHOR_EMAIL", `github\.event\.head_commit\.author\.email`),
+		newUnsafePattern("AUTHOR_NAME", `github\.event\.commits.*?\.author\.name`),
```

Comment:
any reason to not use a map instead of linearly scanning a slice?

Reply:
Ah darn, makes sense though. This may be one of the parts that inspired my questions for you section above (https://github.com/ossf/scorecard/pull/4218#issuecomment-2264066389):

> Any idea how expensive this remediation is? Part of my thoughts with regard to remediation is that there should be some flag to control whether or not it gets surfaced/generated.